work_queue/src/work_queue_server.c

```diff
@@ -0,0 +1,197 @@
+#include "work_queue_json.h"
+#include "link.h"
+#include "jx.h"
+#include "jx_print.h"
+#include "jx_parse.h"
+
+#include <string.h>
+#include <time.h>
+
+char * workqueue = "{ \"name\" : \"server_wq\" , \"port\" : 1234 }";
+
+void reply(struct link *client, char *method, char* message, int id){
+
+    struct jx_pair *jsonrpc, *result, *idd;
+    struct jx *j;
+    char *response;
+    time_t t;
+
+    idd = jx_pair(jx_string("id"), jx_integer(id), NULL);
+
+    if(!strcmp(method, "error")){
+        result = jx_pair(jx_string("error"), jx_string(message), idd);
+    }
+    else {
+        result = jx_pair(jx_string("result"), jx_string(message), idd);
+    }
+        
+    jsonrpc = jx_pair(jx_string("jsonrpc"), jx_string("2.0"), result);
+
+    j = jx_object(jsonrpc);
+
+    response = jx_print_string(j);
+
+    t = time(NULL);
+    link_write(client, response, strlen(response), t + 10);
+
+    jx_delete(j);
+
+}
+
+int main(){
+
+    const char *key;
+    char *error;
+    char message[BUFSIZ];
+    time_t t;
+    void *k = NULL, *v = NULL;
+    char *method, *task;
+    struct jx *val, *value, *jsonrpc;
+    ssize_t read;
+    int timeout, taskid, id=-1;
+
+        //create work queue
+        struct work_queue* q = work_queue_json_create(workqueue);
+    
+        if (!q){
+            printf("Could not create work_queue\n");
+            return 1;
+        }
+ 
+    while(true){
+    
+        //clear work queue
+        work_queue_cancel_all_tasks(q);       
+
+        //wait for client to connect
+        struct link* client = link_serve(2345);
+    
+        if (!client){
+            printf("Could not serve on port 2345\n");
+            return 1;
+        }
+    
+        t = time(NULL);
+    
+        client = link_accept(client, t + 100);
+    
+        if (!client){
+            printf("Could not accept connection\n");
+        }
+    
+        printf("Connected to client. Waiting for messages..\n");
+    
+        while (true){
+    
+            //reset
+            error = NULL;
+            id = -1;
+    
+            //receive message
+            t = time(NULL);
+    
+            read = link_readline(client, message, BUFSIZ, t + 10);
+    
+            if (!read){
+                error = "Error reading from client";
```

Comment:
Under some of these error circumstances, the server should probably drop the connection to the client.  Let's justify those clearly.